command/debug.go

```diff
@@ -0,0 +1,945 @@
+				return err
+			}
+
+			outputLayout[info.Name()] = map[string]interface{}{
+				"timestamp": parsedTime,
+				"files":     []string{},
+			}
+			return nil
+		}
+
+		relPath, err := filepath.Rel(c.flagOutput, path)
+		if err != nil {
+			return err
+		}
+
+		dir, file := filepath.Split(relPath)
+		if len(dir) != 0 {
+			dir = strings.TrimSuffix(dir, "/")
+			filesArr := outputLayout[dir].(map[string]interface{})["files"]
+			outputLayout[dir].(map[string]interface{})["files"] = append(filesArr.([]string), file)
+		} else {
+			outputLayout["files"] = append(outputLayout["files"].([]string), file)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error generating directory output layout: %s", err)
+	}
+
+	c.debugIndex.Output = outputLayout
+
+	// Marshal into json
+	bytes, err := json.MarshalIndent(c.debugIndex, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshaling index file: %s", err)
+	}
+
+	// Write out file
+	if err := ioutil.WriteFile(filepath.Join(c.flagOutput, "index.json"), bytes, 0644); err != nil {
+		return fmt.Errorf("error generating index file; %s", err)
+	}
+
+	return nil
+}
+
+// preflight performs various checks against the provided flags to ensure they
+// are valid/reasonable values. It also takes care of instantiating a client and
+// index object for use by the command.
+func (c *DebugCommand) preflight(rawArgs []string) (*api.Client, string, error) {
+	if !c.skipTimingChecks {
+		// Guard duration and interval values to acceptable values
+		if c.flagDuration < debugMinInterval {
+			c.UI.Info(fmt.Sprintf("Overwriting duration value %q to the minimum value of %q", c.flagDuration, debugMinInterval))
+			c.flagDuration = debugMinInterval
+		}
+		if c.flagInterval < debugMinInterval {
+			c.UI.Info(fmt.Sprintf("Overwriting inteval value %q to the minimum value of %q", c.flagInterval, debugMinInterval))
+			c.flagInterval = debugMinInterval
+		}
+		if c.flagMetricsInterval < debugMinInterval {
+			c.UI.Info(fmt.Sprintf("Overwriting metrics inteval value %q to the minimum value of %q", c.flagMetricsInterval, debugMinInterval))
+			c.flagMetricsInterval = debugMinInterval
+		}
+	}
+
+	// These timing checks are always applicable since interval shouldn't be
+	// greater than the duration
+	if c.flagInterval > c.flagDuration {
+		c.UI.Info(fmt.Sprintf("Overwriting inteval value %q to the duration value %q", c.flagInterval, c.flagDuration))
+		c.flagInterval = c.flagDuration
+	}
+	if c.flagMetricsInterval > c.flagDuration {
+		c.UI.Info(fmt.Sprintf("Overwriting metrics inteval value %q to the duration value %q", c.flagMetricsInterval, c.flagDuration))
+		c.flagMetricsInterval = c.flagDuration
+	}
+
+	if len(c.flagTargets) == 0 {
+		c.flagTargets = c.defaultTargets()
+	}
+
+	// Make sure we can talk to the server
+	client, err := c.Client()
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to create client to connect to Vault: %s", err)
+	}
+	if _, err := client.Sys().Health(); err != nil {
```

Comment:
It's intentional, since debug can technically fetch targets that are unauth'ed (e.g. server-status). All we want to know at this point is whether we're able to connect.

Reply:
I think we're better to fail early if we can detect that we will.  I would either be pessimistic and assume that we'll need a token, or if you want to be cleverer look at what's enabled and decide what endpoint to probe based on that.  My hunch is that most of the time people are going to ask for everything, so we may as well assume that for simplicity.  Unless we think supporting debug-without-a-token is an important use case.

Reply:
Debug-without-a-token is a valid use case, we don't disallow that. I don't think the absence of a token should mean failure here.